Metadata selector fixes and other general cleanup

rfrenchseti opened the pull request:
- Fixed duplicate IDs on left search pane and metadata selector that was causing things the user did in one place to affect the other. Now IDs have a prefix.
- Fixed a bug where when new categories were added or removed from the search tab, the metadata selector wasn't updated.
- In the metadata selector, when a category is expanded scroll so that the new category is visible just like we do on the search tab. Fixes #767.
- In the metadata selector, when a new item it added to the right pane, make sure it is visible.
- When a metadata column is added, make the (i) appear in the correct place relative to ones that were loaded through menu.html.
- When removing a metadata column by clicking on the left pane, the right <li> would fade out, but if you clicked on the trashcan it would just disappear. Now it fades out in both cases.
- Fixed improper font naming of fa-info-circle.
- Recent changes to consolidate code broke what happens when the metadata selector exits with "Save". Now it keeps the slider and the scroll position as it was, rather than resetting back to obs 1. For the table this is basically seamless. For the gallery it will realign the top row with the top of the browser (see issue #761). Fixes #783.
- Make sure the bottom of a recently exposed category is *entirely* on the screen, not just half of it.
- When you click on a table header or sort pill to change the sort order, the table headers compress and then expand (there's no space between the table header label and the sort arrow, then there is). This was caused by a missing class. I've added it and now sorting the table is much smoother.
- General code cleanup.

rfrenchseti committed:
Fix duplicate IDs on search/metadata menus. Fix bug with new categories not showing up in metadata selector. Scroll metadata selector as necessary when expanding categories or adding new metadata. Clean up mutationObserver.

rfrenchseti committed:
Properly label fa-info-circle. Fix <li> contents when adding metadata column. Make removal of column fade out to be consistent with clicking on menu item. Make metadata selector contents scrollbar only adjust on addition of new column.

rfrenchseti committed:
Leave scrollbar where it is when reloading data after selecting metadata.

rfrenchseti committed:
Eliminate annoying table header jump when changing sort order.

rfrenchseti committed:
Fix broken tests

juzen2003 commented on opus/application/static_media/js/mutationObserver.js:
```diff
@@ -37,9 +37,9 @@ var o_mutationObserver = {
-        let adjustSearchWidgetHeight = _.debounce(o_search.adjustSearchHeight, 200);
-        let adjustSearchHeight = _.debounce(o_search.adjustSearchHeight, 200);
+        let searchSideBarHeightChanged = _.debounce(o_search.searchSideBarHeightChanged, 200);
+        let searchWidgetHeightChanged = _.debounce(o_search.searchHeightChanged, 200);
```
This is my bad, it should be a debounced version of original adjustSearchWidgetHeight.
But it's still working fine without issues because adjustSearchWidgetHeight is called inside adjustSearchHeight.
To be more cleaner, it should be like this:
``` js
let searchWidgetHeightChanged = _.debounce(o_search.searchWidgetHeightChanged, 200);
```

rfrenchseti replied:
Fixed

juzen2003 reviewed (commented):
- In tool 8080 server, there is a little white space above the footer in browse tab. But It doesn't exist when running at localhost.
- This is just a note for me, in detail page, when the screen is narrow and image is on top of the left pane, ps can scroll all the way down and get out of the browser window. This issue is fixed in previous responsive branch before merging. I'll look into this later.

rfrenchseti committed:
Fix race condition when user switches to browse tab before we get around to updating the hash and calling result_count

rfrenchseti committed:
Fix type from long ago checkin in mutation observer

rfrenchseti committed:
Fix type from long again checkin in mutation observer

rfrenchseti committed:
Add default value to clearBrowseObservationDataAndEraseDOM.

rfrenchseti committed:
Merge remote-tracking branch 'origin/metadata_selector_767' into metadata_selector_767

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -1947,12 +1956,29 @@ var o_browse = {
         o_browse.observationData = {};
     },
+
+    clearBrowseObservationDataAndEraseDOM: function(leaveStartObs=false) {
```
why don't you need to do this on the cart for similar reasons as well....?

rfrenchseti replied:
clearBrowseObservationDataAndEraseDOM is called from the main OPUS UI when the user changes the search parameters. It makes sure that a data load will be triggered in the browse tab, even if the user switches tabs at an inopportune time (like we didn't actually send out for a result_count yet). But the cart isn't affect by the user's search, so there's no point in clearing it. Since there are now two separate caches for data, there's no overlap between the tabs.

dstopp commented on opus/application/static_media/js/mutationObserver.js:
```diff
@@ -140,18 +139,35 @@ var o_mutationObserver = {
+                        o_browse.adjustMetadataSelectorMenuPS();
+                        // if the collapse opens below the viewable area, move the scrollbar
+                        // only move the scrollbar if we open the menu item
+                        let lastElement = $(mutation.target).children().last();
```
line 186 et all are exactly the same - function?

rfrenchseti replied:
Actually there are quite a few differences - one is doing things to the left side pane (with collapsing categories) and the other is doing things to the right pane (list of selected metadata). Although the code looks similar from a distance, almost every line is different. I looked to see if it would be possible to refactor, but I think it would just make the code more complicated.
